gateway/handler.go

```diff
@@ -352,65 +351,78 @@ func (i *handler) getOrHeadHandler(w http.ResponseWriter, r *http.Request) {
-	if !ok {
+	// TODO: Why did the previous code do path resolution, was that a bug?
+	// TODO: Does If-None-Match apply here?
+	if responseFormat == "application/vnd.ipfs.ipns-record" {
+		logger.Debugw("serving ipns record", "path", contentPath)
+
+		i.addUserHeaders(w)                                 // ok, _now_ write user's headers.
+		w.Header().Set("X-Ipfs-Path", contentPath.String()) // TODO: Should we even set this?
+
+		success := i.serveIpnsRecord(r.Context(), w, r, contentPath, begin, logger)
+
+		if success { // TODO: should we even be reporting this
+			i.getMetric.WithLabelValues(contentPath.Namespace()).Observe(time.Since(begin).Seconds())
+		}
 		return
 	}
-	trace.SpanFromContext(r.Context()).SetAttributes(attribute.String("ResolvedPath", resolvedPath.String()))
 
-	// Detect when If-None-Match HTTP header allows returning HTTP 304 Not Modified
-	if inm := r.Header.Get("If-None-Match"); inm != "" {
-		pathCid := resolvedPath.Cid()
-		// need to check against both File and Dir Etag variants
-		// because this inexpensive check happens before we do any I/O
-		cidEtag := getEtag(r, pathCid)
-		dirEtag := getDirListingEtag(pathCid)
-		if etagMatch(inm, cidEtag, dirEtag) {
-			// Finish early if client already has a matching Etag
-			w.WriteHeader(http.StatusNotModified)
+	var immutableContentPath ImmutablePath
+	if contentPath.Mutable() {
+		immutableContentPath, err = i.api.ResolveMutable(r.Context(), contentPath)
+		if err != nil {
+			// Note: webError will replace http.StatusInternalServerError with a more appropriate error (e.g. StatusNotFound, StatusRequestTimeout, StatusServiceUnavailable, etc.) if necessary
+			err = fmt.Errorf("failed to resolve %s: %w", debugStr(contentPath.String()), err)
+			webError(w, err, http.StatusInternalServerError)
+		}
+	} else {
+		immutableContentPath, err = NewImmutablePath(contentPath)
+		if err != nil {
+			err = fmt.Errorf("path was expected to be immutable, but was not %s: %w", debugStr(contentPath.String()), err)
+			webError(w, err, http.StatusInternalServerError)
 			return
 		}
 	}
 
-	if err := i.handleGettingFirstBlock(r, begin, contentPath, resolvedPath); err != nil {
-		webRequestError(w, err)
+	// Detect when If-None-Match HTTP header allows returning HTTP 304 Not Modified
+	ifNoneMatchResolvedPath, ok := i.handleIfNoneMatch(w, r, responseFormat, contentPath, immutableContentPath, logger)
+	if !ok {
 		return
 	}
 
-	if err := i.setCommonHeaders(w, r, contentPath); err != nil {
-		webRequestError(w, err)
-		return
+	// If we already did the path resolution no need to do it again
+	maybeResolvedImPath := immutableContentPath
+	if ifNoneMatchResolvedPath != nil {
+		maybeResolvedImPath, err = NewImmutablePath(ifNoneMatchResolvedPath)
+		if err != nil {
+			webError(w, err, http.StatusInternalServerError)
+			return
+		}
 	}
 
+	i.addUserHeaders(w) // ok, _now_ write user's headers.
+	w.Header().Set("X-Ipfs-Path", contentPath.String())
```

Comment:
I think we could move these up, so we dont call them twice (once here, and once for `ipns-record`)